### PR TITLE
Generic vehicle turrets

### LINKED ISSUE #8082: Allow turrets to have arbitrary guns attached.
It should be fairly simple to let the player install an arbitrary gun in a turret instead of having separate definitions and recipes for each variety.  We might still want to have different levels of accuracy though,
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/2873863-allow-turrets-to-have-arbitrary-guns-attached?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by irwiss**
#### Summary
Features "Allow mounting any reasonable gun on turret mount"

#### Purpose of change

Fixes #8082

Turrets currently need a separate prototype, this produces tons of boilerplate in e.g. turrets.json that is repetitive and almost identical. Also a lot of guns that would be reasonable as turrets are unavailable for mounting for no reason other than nobody writing the json for it (there is no turret police to stop you from using your .22 pistol in a turret mount after all).

#### Describe the solution

This patch tries to generate the turret prototypes at runtime and derive the correct data for them.

The general idea is scanning all loaded `GUN` itypes, filter out everything not reasonable for a turret and dynamically create turret vparts for them. Some things can be filtered out programmatically (migrations, blacklisted, bionic weapons, pseudo, reload-and-shoot etc).
For guns which can't be filtered out in this way I've added a flag `NO_TURRET` - this prevents the turret vpart from generating ( and if needed have custom json prototype instead ).

Special cases:
* `USE_TANKS` flag - if the gun is using liquid ammo, e.g. flamethrowers, chem sprayers etc, this flag gets added so it uses vehicle tanks.
* `USE_BATTERIES` flag - if the gun is using batteries (that are not UPS), e.g. aftershock battery-magazine stuff get this flag added so they use vehicle batteries.

##### Tilesets

There's a function that approximates how a turret looks according to flags/ammo/ups/skills used and sets looks_like to one of the existing old turret ids of the corresponding type in ultica, this can later be changed so tile ids are just "turret_launcher", "turret_energy" etc instead of specifying per-id tile unless special id is assigned by artist (not yet tuned to cover everything).

##### Install/Remove requirements

Install requirements:
* All turrets require level 3 of the base gun's skill
* Required mechanics level is 3 + (gun weight / 10kg) + (1 if using liquid ammo), capped at 8
* If ups charges are used (energy weapons, electric-fed miniguns etc):
  *  Required electronics level is (ups charges used / 15) capped at 8 rounded up
Removal requirements are the install requirements divided by 2 rounded towards 0

Install time is ( 10 minutes ) * ( gun skill + mechanics required + 2 * ( electronics required ) )
Removal time is half that.

Exact formulas can be fine tuned later, but even simple eyeballing like this produces mostly coherent results.
Time to install/remove makes a bit more sense (even if they can be tuned to fit better ) - M60 now takes 70 mins to install and 35 mins to remove as opposed to the previous 50 seconds.

#### Minor stuff

* pamd68mountable - produces weird name "mounted mountable PA md. 68 battle rifle".

#### Testing

**Warning: loading and saving with this patch will permanently migrate vpart_ids, it's a one way migration so don't use on your "good" saves**

There actually are some unit tests covering vehicle turrets, catching some of the invalid turrets that needed manual filtering out with NO_TURRETS flag.

For generic guns testing [Save file with all vanilla Generic Guns turrets on a bus](https://github.com/CleverRaven/Cataclysm-DDA/files/9190852/generic.guns.-.all.turrets.on.a.bus.zip)

#### Additional context

For reference:
Before patch vanilla has 27 TURRET vparts
After patch vanilla has 291 TURRET vparts
With in-tree mods it ends up 327 TURRET vparts
So rough estimate of 250-ish guns were missing turret prototypes in vanilla.
Expect the numbers to fluctuate around a bit

### CODE REVIEW

**Comment by Night-Pryanik** [comment]
Wow, what else can I say.

File: src/veh_type.cpp [hunk 1]
```diff
@@ -551,10 +551,80 @@ const std::set<std::string> &vpart_info::get_categories() const
...
 void vpart_info::finalize()
 {
     DynamicDataLoader::get_instance().load_deferred( deferred );
 
+    static const auto mountable_gun_filter = []( const itype & t ) {
+        // returns true for "valid" gun items which can be mounted as vehicle turret
+
+        static const std::vector<flag_id> bad_flags {
+            flag_RELOAD_AND_SHOOT,
+            flag_PRIMITIVE_RANGED_WEAPON,
```
Comment (AtomicFox556) [inline src/veh_type.cpp:571]: `PRIMITIVE_RANGED_WEAPON` seems unnecessary here; `RELOAD_AND_SHOOT` would already exclude bows, slings and similar things, while allowing crossbows that would indeed be possible to mount on a vehicle.

Comment (irwiss) [reply src/veh_type.cpp:571]: That makes sense, I'll resolve next time I get to this branch 👍

**Review by Maleclypse — APPROVED** [review]
Aftershock and Xedra Evolved json look good if I’m understanding. Thank you!

**Review by dseguin — APPROVED** [review]

**Comment by sadenar** [comment]
Asking as a translator, will this increase the amount of boilerplate turret strings to actually translate or increase it?

**Comment by irwiss** [comment]
It'll likely be less strings, it adds a string for turret prefix translation, but the names and descriptions are stolen from the gun used for base item, so the few translated turrets are no longer needed

**Comment by sadenar** [comment]
Ah while that's less work, I might have issues with it, because mounted [gun] which was what it was for all those I translated has to be translated, in French at least, by accounting for the gender of the word (machine guns are generally female, pistols male). Will there be a way to individualize these strings again? Don't get me wrong I've done a little bit of tileset work and I won't fight for a good addition to the game to be removed simply because it leads to slightly awkward gendering in one of the 32 loc projects.

**Comment by irwiss** [comment]
I evidently don't know french, can you explain more? Is the "mounted" out of "mounted %s" needs to be gendered for different gun types?
I think it's possible to first check if translated string exists for exact turret part id, and then fallback to the "mounted %s" translation.
You should then be able to default to set "mounted %s" to whatever's correct for majority and fix up individual ids for minorities, does that sound like it'll help or I'm off the mark?

**Pull request merged**